softpack_builder/config/settings.py

```diff
@@ -0,0 +1,145 @@
+"""Copyright (c) Wellcome Sanger Institute.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+"""
+
+import sys
+from pathlib import Path
+from typing import Any, Tuple
+
+import hvac
+import yaml
+from pydantic import BaseSettings
+
+# from pydantic import AnyHttpUrl, AnyUrl, BaseModel, BaseSettings, HttpUrl
+from pydantic.env_settings import SettingsSourceCallable
+
+from softpack_builder.serializable import Serializable
+
+from .models import (
+    ArtifactsConfig,
+    ContainerConfig,
+    EnvironmentsConfig,
+    LoggingConfig,
+    ModulesConfig,
+    ServerConfig,
+    SpackConfig,
+    VaultConfig,
+)
+
+
+class Settings(BaseSettings, Serializable):
+    """Package settings."""
+
+    debug: bool = False
+    server: ServerConfig
+    vault: VaultConfig
+    logging: LoggingConfig
+    spack: SpackConfig
+    environments: EnvironmentsConfig
+    modules: ModulesConfig
+    container: ContainerConfig
+    artifacts: ArtifactsConfig
+
+    class Config:
+        """Configuration loader."""
+
+        config_dir = "conf"
+        config_file = "config.yml"
+
+        @classmethod
+        def file_settings(
+            cls, path: Path, settings: BaseSettings
+        ) -> dict[str, Any]:
+            """Load settings from file.
+
+            Args:
+                path: Config file path.
+                settings: Base settings object.
+
+            Returns:
+                dict[str, Any]: A dictionary of settings.
+            """
+            if not path.is_file():
+                return {}
+            with open(path) as f:
+                return yaml.safe_load(f)
+
+        @classmethod
+        def defaults(cls, settings: BaseSettings) -> dict[str, Any]:
+            """Load defaults from config file.
+
+            Args:
+                settings: BaseSettings model.
+
+            Returns:
+               dict[str, Any]: Settings loaded from default config file.
+            """
+            package_dir = Path(__file__).parent.absolute()
+            path = package_dir / cls.config_dir / cls.config_file
+            return cls.file_settings(path, settings)
+
+        @classmethod
+        def overrides(cls, settings: BaseSettings) -> dict[str, Any]:
+            """Load overrides from config file in the home directory.
+
+            Args:
+                settings: BaseSettings model.
+
+            Returns:
+                dict[str, Any]: Settings loaded from deployment-specific
+                config file.
+            """
+            path = Path.home() / ".softpack/builder" / cls.config_file
```

Comment:
Maybe ".softpack/builder" counts as a magic value, and should be assigned to a variable therefore

Reply:
fixed